### PR TITLE
Fix build_boundary_part for touching rings

### PR DESCRIPTION
**Pull request by amatissart**
This is based on an old branch I never took the time to clean up.  
It solves a subtle issue about boundaries composed of multiple outer rings that touch at a single point.  
For example in Austria: https://www.openstreetmap.org/relation/16239#map=14/47.5656/10.4596

Depending on the order of the ways in the OSM relation, in some cases only 1 ring was built with duplicated nodes in it, whereas two distinct polygons are expected.

This new implementation keeps track of the added nodes while iterating on the boundary parts, to extract minimal rings.

### CODE REVIEW

File: src/boundaries.rs [hunk 1]
```diff
@@ -5,17 +5,17 @@ use std::borrow::Borrow;
...
+use crate::osm_builder::named_node;
 
 struct BoundaryPart {
     nodes: Vec<osmpbfreader::Node>,
 }
 
 impl BoundaryPart {
     pub fn new(nodes: Vec<osmpbfreader::Node>) -> BoundaryPart {
-        BoundaryPart { nodes: nodes }
+        BoundaryPart { nodes }
```
Comment (remi-dupre) [inline src/boundaries.rs:20]: As `BoundaryPart`'s role is to be a wrapper around a non-empty array it maybe cleaner to check that it is actually non-empty here? Maybe return an `Option` to explicit the possibility of failure and use this option to replace external logic over the size of the array?

*Maybe this should be left for another PR though?*

File: src/boundaries.rs [hunk 2]
```diff
@@ -163,48 +163,69 @@ pub fn build_boundary_parts<T: Borrow<osmpbfreader::OsmObj>>(
...
+                .iter()
+                .map(|n| Coordinate {
+                    x: n.lon(),
+                    y: n.lat(),
+                })
+                .collect();
+            multipoly
+                .0
+                .push(Polygon::new(LineString(poly_geom), vec![]));
+        };
```
Comment (remi-dupre) [inline src/boundaries.rs]: It would probably be clearer to define this helper function outside of the while loop, probably right next the definition of `multipoly` as this is the only context it depends on?

File: src/boundaries.rs [hunk 3]
```diff
@@ -213,17 +234,20 @@ pub fn build_boundary_parts<T: Borrow<osmpbfreader::OsmObj>>(
...
-                        relation.tags.get("name").map_or("", |s| &s),
-                        poly_geom.first().unwrap().id,
-                        poly_geom.last().unwrap().id,
-                        distance
-                    );
+
+                if added_nodes.len() > 1 {
+                    let distance = p(added_nodes.first().unwrap())
+                        .haversine_distance(&p(added_nodes.last().unwrap()));
+                    if distance < 10. {
```
Comment (remi-dupre) [inline src/boundaries.rs]: It may be better to use `10.` as a previously named constant? Something like `MAX_CYCLE_CLOSING_DISTANCE`?

*Maybe this should be left for another PR though?*

**Review by remi-dupre — APPROVED** [review]
The algorithm looks solid, I mostly added comments about unaltered parts of the existing code. I'll probably open another PR addressing this kind of things if you also think it sounds relevant?

While reading `boundaries.rs` I was wondering if we could actually simplify the algorithmic part by flattening all the ways into a simple undirected and unweighted graph?

File: src/boundaries.rs (hunk 2 above)
Comment (amatissart) [reply src/boundaries.rs]: done

File: src/boundaries.rs (hunk 3 above)
Comment (amatissart) [reply src/boundaries.rs]: done
